src/main.cpp

```diff
@@ -4155,26 +4156,58 @@ void static ProcessGetData(CNode* pfrom, const Consensus::Params& consensusParam
     }
 }
 
-bool static ProcessMessage(CNode* pfrom, std::string strCommand, CDataStream& vRecv, int64_t nTimeReceived)
-{
-    const CChainParams& chainparams = Params();
-    RandAddSeedPerfmon();
-    const bool fReindex = Blocks::DB::instance()->isReindexing();
-    logDebug(Log::Net) << "received:" << SanitizeString(strCommand) << "bytes:" << vRecv.size() << "peer:" << pfrom->id;
-    if (mapArgs.count("-dropmessagestest") && GetRand(atoi(mapArgs["-dropmessagestest"])) == 0)
-    {
-        LogPrintf("dropmessagestest DROPPING RECV MESSAGE\n");
-        return true;
-    }
+class NetworkMessage {
+public:
+    virtual ~NetworkMessage() {}
+
+    NetworkMessage(CNode * const pFrom,
```

Comment:
I think one constructor with only the pFrom should be fine.

More constructors and you just create problems for yourself because all the inheriting classes should mirror that.

The chainParams is a global singleton, I think we can stop passing it along in all of Bitcoin Classic.

Maybe the easiest solution for the booleans is to have some setters and make the constructor just default them to off.

Reply:
almost done with the functional refactor (couldn't help myself and code this)

just to be on the safe side, I'm receiving the `chainParams` parameter. Not that sure it's a global singleton as I see it being instantiated in many different places

![screen shot 2017-09-20 at 7 20 28 am](https://user-images.githubusercontent.com/163977/30645859-45568f74-9dd4-11e7-9a45-85dbc895d1fd.png)

although internally `chainparams.cpp` it returns the same pointer when the global `Params()` function is invoked.

![screen shot 2017-09-20 at 7 21 51 am](https://user-images.githubusercontent.com/163977/30645878-5bd774fc-9dd4-11e7-80ac-e04a82faf5ed.png)

So another option I guess, is to not put that parameter in the method interface and just invoke `Params()` internally, but I like having the flexibility of passing any `CCHainParams` pointer, specially for testing.

Reply:
The idea of having that flexibility in testing is however turns out to not be used as the testing framework sets the proper Params in the start of each test already and experience shows we don't want to change it halfway through a single test.

See BasicTestingSetup constructor where this is done.

Reply:
ok, makes sense to get rid of this param and use `Params()` like I've seen you've been doing in recent fixes. 

Making this move now to `messages/` is letting me see what a clusterfuck there is in main.h and main.cpp.